tests/Compiler/Type/PhpDocTypeUtilsTest.php

```diff
@@ -417,4 +422,404 @@ public function testResolve(): void
+    {
+        $typeNode = $this->parseType($type);
+
+        foreach ($true as $trueType) {
+            $trueTypeNode = $this->parseType($trueType);
+            self::assertTrue(PhpDocTypeUtils::isSubTypeOf($trueTypeNode, $typeNode), "$trueType is not subtype of $type");
+        }
+
+        foreach ($false as $falseType) {
+            $falseTypeNode = $this->parseType($falseType);
+            self::assertFalse(PhpDocTypeUtils::isSubTypeOf($falseTypeNode, $typeNode), "$falseType is subtype of $type");
+        }
+    }
+
+    private function parseType(string $type): TypeNode
+    {
+        $lexer = new Lexer();
+        $constExprParser = new ConstExprParser(unescapeStrings: true);
+        $typeParser = new TypeParser($constExprParser);
+
+        return $typeParser->parse(new TokenIterator($lexer->tokenize($type)));
+    }
+
+    /**
+     * @return iterable<string, array{type: string, true: list<string>, false: list<string>}>
+     */
+    public static function provideIsSubTypeOfData(): iterable
+    {
+        yield 'array' => [
+            'type' => 'array',
+
+            'true' => [
+                'array',
+                'int[]',
+                'array{int}',
+                'list<int>',
+                'array<int>',
+                'array<int, int>',
+            ],
+
+            'false' => [
+                'int',
+                'iterable',
+            ],
+        ];
+
+        yield 'array<int>' => [
+            'type' => 'array<int>',
+
+            'true' => [
+                'array<int>',
+                'array<int, int>',
+                'list<int>',
+                'array{int}',
+                'int[]',
+            ],
+
+            'false' => [
+                'array',
+                'array{int, string}',
+                'int',
+            ],
+        ];
+
+        yield 'array<string, int>' => [
+            'type' => 'array<string, int>',
+
+            'true' => [
+                'array<string, int>',
+                'array{foo: int}',
+                'array{"foo": int}',
+            ],
+
+            'false' => [
+                'array',
+                'array<int, string>',
+                'array{string}',
+                'array{123: string}',
+                'int',
+            ],
+        ];
+
+        yield 'bool' => [
+            'type' => 'bool',
+
+            'true' => [
+                'bool',
+                'boolean',
+                'true',
+                'false',
+            ],
+
+            'false' => [
+                'int',
+            ],
+        ];
+
+        yield 'callable' => [
+            'type' => 'callable',
+
+            'true' => [
+                'callable',
+                'callable(): int',
+                'Closure',
+                'Closure(): int',
+                '"strval"',
+            ],
+
+            'false' => [
+                'int',
+                '"abc"',
+            ],
+        ];
```

Comment:
Can we test class with __invoke?

Reply:
`Closure` is a class with `__invoke`. And I don't think there is any other built-in class, that is callable.